// creem-convex/dist/index.js
import { defineTable } from 'convex/server';
import { v } from 'convex/values';

// src/client.ts
var PROD_URL = "https://api.creem.io";
var TEST_URL = "https://test-api.creem.io";
var CreemClient = class {
  constructor(config) {
    this.apiKey = config.apiKey;
    this.baseUrl = config.testMode ? TEST_URL : PROD_URL;
  }
  async request(method, path, body) {
    const url = `${this.baseUrl}${path}`;
    const headers = {
      "x-api-key": this.apiKey,
      "Content-Type": "application/json"
    };
    const res = await fetch(url, {
      method,
      headers,
      ...body ? { body: JSON.stringify(body) } : {}
    });
    if (!res.ok) {
      const errBody = await res.json().catch(() => ({ message: res.statusText }));
      throw new CreemApiError(
        res.status,
        errBody.message ?? res.statusText,
        errBody.trace_id
      );
    }
    return res.json();
  }
  // --- Products ---
  async getProduct(productId) {
    return this.request("GET", `/v1/products?id=${productId}`);
  }
  async listProducts(page = 1, limit = 10) {
    return this.request("GET", `/v1/products/search?page=${page}&limit=${limit}`);
  }
  async createProduct(params) {
    return this.request("POST", "/v1/products", {
      name: params.name,
      description: params.description,
      price: params.price,
      currency: params.currency ?? "USD",
      billing_type: params.billingType,
      billing_period: params.billingPeriod
    });
  }
  // --- Checkouts ---
  async createCheckout(params) {
    return this.request("POST", "/v1/checkouts", {
      product_id: params.productId,
      success_url: params.successUrl,
      request_id: params.requestId,
      units: params.units,
      discount_code: params.discountCode,
      customer: params.customer,
      metadata: params.metadata,
      custom_fields: params.customFields
    });
  }
  async getCheckout(checkoutId) {
    return this.request("GET", `/v1/checkouts?id=${checkoutId}`);
  }
  // --- Subscriptions ---
  async getSubscription(subscriptionId) {
    return this.request("GET", `/v1/subscriptions?id=${subscriptionId}`);
  }
  async cancelSubscription(subscriptionId) {
    return this.request("POST", `/v1/subscriptions/${subscriptionId}/cancel`);
  }
  async pauseSubscription(subscriptionId) {
    return this.request("POST", `/v1/subscriptions/${subscriptionId}/pause`);
  }
  async resumeSubscription(subscriptionId) {
    return this.request("POST", `/v1/subscriptions/${subscriptionId}/resume`);
  }
  // --- Customers ---
  async getCustomer(customerId) {
    return this.request("GET", `/v1/customers?id=${customerId}`);
  }
  async getCustomerByEmail(email) {
    return this.request("GET", `/v1/customers?email=${encodeURIComponent(email)}`);
  }
  async listCustomers(page = 1, limit = 10) {
    return this.request("GET", `/v1/customers/list?page=${page}&limit=${limit}`);
  }
  async createBillingPortal(customerId) {
    return this.request("POST", "/v1/customers/billing", {
      customer_id: customerId
    });
  }
};
var CreemApiError = class extends Error {
  constructor(status, message, traceId) {
    super(Array.isArray(message) ? message.join(", ") : message);
    this.status = status;
    this.traceId = traceId;
    this.name = "CreemApiError";
  }
};
var creemTables = {
  creem_payments: defineTable({
    creemCheckoutId: v.string(),
    creemOrderId: v.optional(v.string()),
    creemCustomerId: v.optional(v.string()),
    creemProductId: v.string(),
    creemSubscriptionId: v.optional(v.string()),
    amount: v.number(),
    currency: v.string(),
    status: v.string(),
    // "paid" | "pending" | "failed"
    type: v.string(),
    // "one_time" | "recurring"
    customerEmail: v.optional(v.string()),
    customerName: v.optional(v.string()),
    metadata: v.optional(v.any()),
    creemCreatedAt: v.string()
  }).index("by_checkout", ["creemCheckoutId"]).index("by_customer", ["creemCustomerId"]).index("by_product", ["creemProductId"]).index("by_subscription", ["creemSubscriptionId"]),
  creem_subscriptions: defineTable({
    creemSubscriptionId: v.string(),
    creemCustomerId: v.string(),
    creemProductId: v.string(),
    status: v.string(),
    // CreemSubscriptionStatus
    customerEmail: v.optional(v.string()),
    customerName: v.optional(v.string()),
    currentPeriodStart: v.optional(v.string()),
    currentPeriodEnd: v.optional(v.string()),
    canceledAt: v.optional(v.string()),
    metadata: v.optional(v.any()),
    creemCreatedAt: v.string(),
    creemUpdatedAt: v.string()
  }).index("by_creem_id", ["creemSubscriptionId"]).index("by_customer", ["creemCustomerId"]).index("by_product", ["creemProductId"]).index("by_status", ["status"]),
  creem_customers: defineTable({
    creemCustomerId: v.string(),
    email: v.string(),
    name: v.optional(v.string()),
    country: v.optional(v.string()),
    creemCreatedAt: v.string()
  }).index("by_creem_id", ["creemCustomerId"]).index("by_email", ["email"]),
  creem_webhook_events: defineTable({
    creemEventId: v.string(),
    eventType: v.string(),
    processed: v.boolean(),
    payload: v.any(),
    error: v.optional(v.string()),
    creemCreatedAt: v.number()
    // timestamp
  }).index("by_event_id", ["creemEventId"]).index("by_type", ["eventType"])
};

// src/webhook/verify.ts
async function verifyWebhookSignature(payload, signature, secret) {
  const encoder = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  const sig = await crypto.subtle.sign("HMAC", key, encoder.encode(payload));
  const computed = Array.from(new Uint8Array(sig)).map((b) => b.toString(16).padStart(2, "0")).join("");
  return computed === signature;
}

// src/webhook/handler.ts
var GRANT_EVENTS = [
  "subscription.active",
  "subscription.trialing",
  "subscription.paid"
];
var REVOKE_EVENTS = [
  "subscription.canceled",
  "subscription.expired",
  "subscription.paused"
];
async function handleCreemWebhook(ctx, request, handlers) {
  const signature = request.headers.get("creem-signature");
  if (!signature) {
    return new Response("Missing creem-signature header", { status: 401 });
  }
  const rawBody = await request.text();
  const webhookSecret = process.env.CREEM_WEBHOOK_SECRET;
  if (!webhookSecret) {
    console.error("CREEM_WEBHOOK_SECRET environment variable not set");
    return new Response("Server configuration error", { status: 500 });
  }
  const isValid = await verifyWebhookSignature(rawBody, signature, webhookSecret);
  if (!isValid) {
    return new Response("Invalid signature", { status: 401 });
  }
  let event;
  try {
    event = JSON.parse(rawBody);
  } catch {
    return new Response("Invalid JSON", { status: 400 });
  }
  const webhookCtx = {
    runMutation: ctx.runMutation.bind(ctx),
    runQuery: ctx.runQuery.bind(ctx),
    runAction: ctx.runAction.bind(ctx)
  };
  try {
    await routeEvent(event, handlers, webhookCtx);
  } catch (err) {
    console.error(`Error handling ${event.eventType}:`, err);
    return new Response("Handler error", { status: 500 });
  }
  return new Response(JSON.stringify({ received: true }), {
    status: 200,
    headers: { "Content-Type": "application/json" }
  });
}
async function routeEvent(event, handlers, ctx) {
  const { eventType } = event;
  const handlerMap = {
    "checkout.completed": handlers.onCheckoutCompleted,
    "subscription.active": handlers.onSubscriptionActive,
    "subscription.paid": handlers.onSubscriptionPaid,
    "subscription.canceled": handlers.onSubscriptionCanceled,
    "subscription.expired": handlers.onSubscriptionExpired,
    "subscription.trialing": handlers.onSubscriptionTrialing,
    "subscription.paused": handlers.onSubscriptionPaused,
    "subscription.past_due": handlers.onSubscriptionPastDue,
    "subscription.update": handlers.onSubscriptionUpdate,
    "subscription.scheduled_cancel": handlers.onSubscriptionScheduledCancel,
    "refund.created": handlers.onRefundCreated,
    "dispute.created": handlers.onDisputeCreated
  };
  const handler = handlerMap[eventType];
  if (handler) {
    await handler(ctx, event.object);
  }
  if (GRANT_EVENTS.includes(eventType) && handlers.onGrantAccess) {
    await handlers.onGrantAccess(ctx, event.object);
  }
  if (REVOKE_EVENTS.includes(eventType) && handlers.onRevokeAccess) {
    await handlers.onRevokeAccess(ctx, event.object);
  }
}

// src/actions/index.ts
function getCreemClient() {
  const apiKey = process.env.CREEM_API_KEY;
  if (!apiKey) throw new Error("CREEM_API_KEY environment variable not set");
  const testMode = process.env.CREEM_TEST_MODE === "true";
  return new CreemClient({ apiKey, testMode });
}
async function creemCreateCheckout(params) {
  const client = getCreemClient();
  const checkout = await client.createCheckout(params);
  return {
    checkoutId: checkout.id,
    checkoutUrl: checkout.checkout_url,
    status: checkout.status
  };
}
async function creemGetProduct(productId) {
  const client = getCreemClient();
  return await client.getProduct(productId);
}
async function creemListProducts(page = 1, limit = 10) {
  const client = getCreemClient();
  return await client.listProducts(page, limit);
}
async function creemGetSubscription(subscriptionId) {
  const client = getCreemClient();
  return await client.getSubscription(subscriptionId);
}
async function creemCancelSubscription(subscriptionId) {
  const client = getCreemClient();
  return await client.cancelSubscription(subscriptionId);
}
async function creemPauseSubscription(subscriptionId) {
  const client = getCreemClient();
  return await client.pauseSubscription(subscriptionId);
}
async function creemResumeSubscription(subscriptionId) {
  const client = getCreemClient();
  return await client.resumeSubscription(subscriptionId);
}
async function creemGetCustomer(opts) {
  const client = getCreemClient();
  if (opts.customerId) return await client.getCustomer(opts.customerId);
  if (opts.email) return await client.getCustomerByEmail(opts.email);
  throw new Error("Either customerId or email is required");
}
async function creemCreateBillingPortal(customerId) {
  const client = getCreemClient();
  return await client.createBillingPortal(customerId);
}

// src/mutations/index.ts
async function syncCheckoutToDb(db, checkout) {
  const order = checkout.order;
  const customer = checkout.customer;
  const product = checkout.product;
  if (customer) {
    await upsertCustomer(db, customer);
  }
  await db.insert("creem_payments", {
    creemCheckoutId: checkout.id,
    creemOrderId: order?.id,
    creemCustomerId: customer?.id,
    creemProductId: product?.id ?? "",
    creemSubscriptionId: checkout.subscription?.id,
    amount: order?.amount ?? 0,
    currency: order?.currency ?? "USD",
    status: order?.status ?? "paid",
    type: order?.type ?? "one_time",
    customerEmail: customer?.email,
    customerName: customer?.name,
    metadata: checkout.metadata,
    creemCreatedAt: order?.created_at ?? (/* @__PURE__ */ new Date()).toISOString()
  });
  if (checkout.subscription) {
    await upsertSubscription(db, checkout.subscription, customer);
  }
}
async function syncSubscriptionToDb(db, subscription) {
  const customer = typeof subscription.customer === "object" ? subscription.customer : null;
  await upsertSubscription(db, subscription, customer);
}
async function upsertCustomer(db, customer) {
  const existing = await db.query("creem_customers").withIndex("by_creem_id", (q) => q.eq("creemCustomerId", customer.id)).unique();
  const data = {
    creemCustomerId: customer.id,
    email: customer.email,
    name: customer.name,
    country: customer.country,
    creemCreatedAt: customer.created_at
  };
  if (existing) {
    await db.patch(existing._id, data);
  } else {
    await db.insert("creem_customers", data);
  }
}
async function upsertSubscription(db, sub, customer) {
  const existing = await db.query("creem_subscriptions").withIndex("by_creem_id", (q) => q.eq("creemSubscriptionId", sub.id)).unique();
  const productId = typeof sub.product === "object" ? sub.product.id : sub.product;
  const customerId = typeof sub.customer === "object" ? sub.customer.id : sub.customer;
  const data = {
    creemSubscriptionId: sub.id,
    creemCustomerId: customerId,
    creemProductId: productId,
    status: sub.status,
    customerEmail: customer?.email,
    customerName: customer?.name,
    currentPeriodStart: sub.current_period_start_date ?? void 0,
    currentPeriodEnd: sub.current_period_end_date ?? void 0,
    canceledAt: sub.canceled_at ?? void 0,
    metadata: sub.metadata,
    creemCreatedAt: sub.created_at,
    creemUpdatedAt: sub.updated_at
  };
  if (existing) {
    await db.patch(existing._id, data);
  } else {
    await db.insert("creem_subscriptions", data);
  }
}
async function logWebhookEvent(db, eventId, eventType, payload, createdAt) {
  const existing = await db.query("creem_webhook_events").withIndex("by_event_id", (q) => q.eq("creemEventId", eventId)).unique();
  if (existing) return existing._id;
  return await db.insert("creem_webhook_events", {
    creemEventId: eventId,
    eventType,
    processed: true,
    payload,
    creemCreatedAt: createdAt
  });
}

// src/queries/index.ts
async function querySubscriptionByCustomer(db, creemCustomerId) {
  return await db.query("creem_subscriptions").withIndex("by_customer", (q) => q.eq("creemCustomerId", creemCustomerId)).order("desc").first();
}
async function querySubscriptionById(db, creemSubscriptionId) {
  return await db.query("creem_subscriptions").withIndex("by_creem_id", (q) => q.eq("creemSubscriptionId", creemSubscriptionId)).unique();
}
async function queryActiveSubscriptions(db) {
  return await db.query("creem_subscriptions").withIndex("by_status", (q) => q.eq("status", "active")).collect();
}
async function querySubscriptionsByProduct(db, creemProductId) {
  return await db.query("creem_subscriptions").withIndex("by_product", (q) => q.eq("creemProductId", creemProductId)).collect();
}
async function queryPaymentsByCustomer(db, creemCustomerId) {
  return await db.query("creem_payments").withIndex("by_customer", (q) => q.eq("creemCustomerId", creemCustomerId)).order("desc").collect();
}
async function queryCustomerByEmail(db, email) {
  return await db.query("creem_customers").withIndex("by_email", (q) => q.eq("email", email)).unique();
}
async function queryHasActiveSubscription(db, creemCustomerId) {
  const sub = await db.query("creem_subscriptions").withIndex("by_customer", (q) => q.eq("creemCustomerId", creemCustomerId)).order("desc").first();
  if (!sub) return { hasAccess: false, status: null, subscription: null };
  const activeStatuses = ["active", "trialing"];
  return {
    hasAccess: activeStatuses.includes(sub.status),
    status: sub.status,
    subscription: sub
  };
}
async function queryWebhookEvents(db, limit = 50) {
  return await db.query("creem_webhook_events").order("desc").take(limit);
}

export { CreemApiError, CreemClient, creemCancelSubscription, creemCreateBillingPortal, creemCreateCheckout, creemGetCustomer, creemGetProduct, creemGetSubscription, creemListProducts, creemPauseSubscription, creemResumeSubscription, creemTables, getCreemClient, handleCreemWebhook, logWebhookEvent, queryActiveSubscriptions, queryCustomerByEmail, queryHasActiveSubscription, queryPaymentsByCustomer, querySubscriptionByCustomer, querySubscriptionById, querySubscriptionsByProduct, queryWebhookEvents, syncCheckoutToDb, syncSubscriptionToDb, upsertCustomer, upsertSubscription, verifyWebhookSignature };
